### ISSUE TITLE
Core/SAI: SMART_EVENT_UPDATE_OOC seems to ignore %-chance

### ISSUE THREAD

**Issue by Jinnaix** [TrinityCore/TrinityCore#25209]
**Description:**

While i was messing around with [Scarshield Portal](https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=9707#abilities) script by removing the despawn, i noticed that OOC completly ignores the percentage chance for casts. The Burning Felguards have been summoned in every lifetime of the portal.
Might be the case that it completely ignores event conditions.

**Expected behaviour:**

Non 100% chances should be recognized by SMART_EVENT_UPDATE_OOC

**Steps to reproduce the problem:**

1. 
```sql
DELETE FROM `smart_scripts` WHERE (`entryorguid` = 9707) AND (`source_type` = 0) AND (`id` IN (3));
``` 
This is to prevent the Portal form despawn itself after 10 seconds
2. go c id 9257 and get in combat with the warlocks
3. Watch

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

174ac958009222aeef83520101a28963f8c63b0e

**Operating system:** Win10

**Comment by offl** [comment]
For some reason it works with another actions, for example:
use 10-20 times .npc add temp 31397
after 10-30 seconds only few of them will say text (15% chance to say)

Definitely the chance is ignored in case of SPELL CAST action:
use .npc add temp 18037
step back (to spawn NPC in different place) and use it again
repeat 10-20 times
notice every time creature summons imp, but the chance to summon is only 50%

The chance is ignored also with this script (apply it)
```sql
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 9707);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(9707, 0, 0, 0, 1, 0, 100, 0, 5000, 5000, 5000, 5000, 0, 11, 15126, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Scarshield Portal - Out of Combat - Cast \'Summon Burning Imp\' (Normal Dungeon)'),
(9707, 0, 1, 0, 1, 0, 100, 0, 11000, 11000, 11000, 11000, 0, 11, 16002, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Scarshield Portal - Out of Combat - Cast \'Summon Burning Felhound\' (Normal Dungeon)'),
(9707, 0, 2, 0, 1, 0, 10, 1, 20000, 20000, 0, 0, 0, 11, 16004, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Scarshield Portal - Out of Combat - Cast \'Summon Burning Felguard\' (Normal Dungeon)'),
(9707, 0, 4, 0, 63, 0, 100, 0, 0, 0, 0, 0, 0, 103, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'Scarshield Portal - On Just Created - Set Rooted On');
```
.go c id 9257
aggro, let creature spawn portal and wait 20 sec
repeat 3 times, every time Burning Felguard is spawned while the chance to cast spell is only 10%

**I checked one spell cast in combat and the chance was not ignored.**
.npc add temp 25367
aggro and wait 5 sec for frost nova
the chance is not ignores since sometimes creature does not cast it (use .npc add temp 25367 again and again and don't forget to kill previously spawned creature)

**Comment by jackpoz** [comment]
The chance does work, the issue is that it's retried over and over, with the timer always set at 0 so "ready to be executed". You can verify this by changing the chance of npc 18037 from 50% to 1% and see how after some time it spawns the imp.

This is the code handling the run once flag:
```cpp
(e.event.event_flags & SMART_EVENT_FLAG_NOT_REPEATABLE) && e.runOnce
```
and this is the code rolling the chance:
https://github.com/TrinityCore/TrinityCore/blob/0ad1854d1c6a3b6ff93e71272c919f1f5f11d5ed/src/server/game/AI/SmartScripts/SmartScript.cpp#L247-L255

Notice how runOnce is set to true only after a successful chance. That means the event will be retried over and over until it finally succeeds.

Should runOnce be set even if the roll fails ? I would say that would make more sense.

**Comment by offl** [comment]
> the issue is that it's retried over and over, with the timer always set at 0 so "ready to be executed"

Well, as I told you this should not happen because there are non-repeatable actions with linked to them actions like say, play emote and other stuff. If for some reason creature can't cast spell, it will repeat all other actions over and over again(currently happening)
Action should be "holded" until creature can execute them. Without trying to execute them every 50ms because there is difference between "check if can execute" and "try to execute". You can "check if can execute" by checking if creature already casts spell and can cast this specific spell because I guess some spells ARE supposed to be casted even if creature already casts another spell

> Should runOnce be set even if the roll fails ?

If roll fails - definitely yes. For example we want to cast spell  with 5% chance, only once. Event has no repeat timers. So if roll wasn't sucessful we don't want to try to execute it again and again
But if roll didn't failed, we should try to cast spell when we can and set runOnce after sucessful cast

**Comment by jackpoz** [comment]
> > the issue is that it's retried over and over, with the timer always set at 0 so "ready to be executed"
> 
> Well, as I told you this should not happen because there are non-repeatable actions with linked to them actions like say, play emote and other stuff. If for some reason creature can't cast spell, it will repeat all other actions over and over again(currently happening)
> Action should be "holded" until creature can execute them. Without trying to execute them every 50ms because there is difference between "check if can execute" and "try to execute". You can "check if can execute" by checking if creature already casts spell and can cast this specific spell because I guess some spells ARE supposed to be casted even if creature already casts another spell

This has nothing to do with this case. SmartScript::ProcessAction() rolls the chance and if the chance fails the action will be retried again.



> > Should runOnce be set even if the roll fails ?
> 
> But if roll didn't failed, we should try to cast spell when we can and set runOnce after sucessful cast

This is another issue specifically related to actions that can fail and should be retried. Better fix that in a separate fix.

**Comment by jackpoz** [comment]
I would like to test the case of an action that can be repeated every X seconds with Y % chance and seeing ingame that sometimes it works, sometimes it doesn't, with the right X cooldown respected

**Comment by jackpoz** [comment]
I created a PR at https://github.com/TrinityCore/TrinityCore/pull/25778 . Binaries can be downloaded from appveyor if one wants to test it without rebuilding

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/commit/96b289cadbf304e427ddbd281d1a6f592af612a7
